### ISSUE TITLE
What's the exactly formula of `softmax_lse`?

### ISSUE THREAD

**Issue by flymark2010** [Dao-AILab/flash-attention#404]
Hi @tridao 
I want to use the value of [softmax_lse](https://github.com/Dao-AILab/flash-attention/blob/184b992dcb2a0890adaa19eb9b541c3e4f9d2a08/flash_attn/flash_attn_interface.py#L482C4-L482C4). What's the exactly formula, $\log\sum\exp(x_i)$ or $\log\sum\exp(x_i - \max(x_i))$ ?

**Comment by tridao** [comment]
log sum exp(x_i). Note that softmax_lse is only meant to be used internally so e.g. the gradient of softmax_lse won't be taken into account during the backward pass.

**Comment by flymark2010** [comment]
Got it, thanks for reminding!

**Comment by shadow150519** [comment]
@tridao @flymark2010 Hi, I want to confirm this again: softmax lse should directly return `log sum exp(x_i)` rather than `log sum exp(x_i - x_max_block_i)` when split kv into kv blocks. Assuming one q block and  two kv blocks, I'll get o1, o2, lse1, lse2. When deriving online softmax, I found that if lse returns `log sum exp(x_i - x_max_block_i)`, I seem unable to compute the correct final o and lse, because each block's output is scaled by its local max. Only by directly returning log sum exp(x_i) can we merge the results correctly.

**Comment by zhengqigao** [comment]
@shadow150519 I happened to do some testing. Please see the following comparison with a reference PyTorch implementation, which confirms the meaning of `lse`. 

```python
import flash_attn
from flash_attn import flash_attn_varlen_func
from flash_attn.flash_attn_interface import (
    _flash_attn_varlen_forward,
    _flash_attn_varlen_backward,
)
flash_attn_version = getattr(flash_attn, "__version__", None)
import torch


def reference_attention(q, k, v, cu_seqlens, causal=True):
    """
    q: [total_tokens, nhead, headdim]
    k: [total_tokens, nhead, headdim]
    v: [total_tokens, nhead, headdim]
    cu_seqlens: [batch+1] 
    Returns: attention output same shape as q, plus (optionally) lse
    """
    device = q.device
    batch_size = cu_seqlens.numel() - 1
    nhead = q.shape[1]
    headdim = q.shape[2]
    out = torch.empty_like(q)
    lse = torch.empty(nhead, q.shape[0], device=q.device, dtype=torch.float32)

    for b in range(batch_size):
        start = cu_seqlens[b].item()
        end = cu_seqlens[b + 1].item()
        q_seq = q[start:end]          # [seq, nhead, headdim]
        k_seq = k[start:end]          # [seq, nhead, headdim]
        v_seq = v[start:end]          # [seq, nhead, headdim]
        seq_len = end - start
        # [nhead, seq, headdim] (for batch/sequence b)
        q_b = q_seq.transpose(0, 1)   # [nhead, seq, headdim]
        k_b = k_seq.transpose(0, 1)   # [nhead, seq, headdim]
        v_b = v_seq.transpose(0, 1)   # [nhead, seq, headdim]
        attn_scores = torch.matmul(q_b.float(), k_b.transpose(-1, -2).float()) # [nhead, seq, seq]
        attn_scores = attn_scores / (headdim ** 0.5)
        if causal:
            mask = torch.triu(torch.ones(seq_len, seq_len, device=device), diagonal=1)
            attn_scores = attn_scores.masked_fill(mask.bool().unsqueeze(0), float('-inf'))
        attn_weights = torch.softmax(attn_scores, dim=-1) # [nhead, seq, seq]
        out_b = torch.matmul(attn_weights.float(), v_b.float()) # [nhead, seq, headdim]
        out_seq = out_b.transpose(0, 1) # [seq, nhead, headdim]
        out[start:end] = out_seq.to(dtype=q.dtype)
        # compute logsumexp for each q position and head
        lse_b = (torch.logsumexp(attn_scores, dim=-1) ) # [nhead, seq]
        lse[:,start:end] = lse_b
    return out, lse


def test_flash_attn_varlen_three_seq():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    dtype = torch.bfloat16

    # 3 sequences: lengths 3, 6, 5
    cu_seqlens = torch.tensor([0, 3, 9, 14], dtype=torch.int32, device=device)
    max_seqlen = int((cu_seqlens[1:] - cu_seqlens[:-1]).max().item())
    seqlen = cu_seqlens[-1].item()
    nhead = 4
    headdim = 8

    # Construct qkv: [total_tokens, nhead, headdim]
    q = torch.randn(seqlen, nhead, headdim, device=device, dtype=dtype).requires_grad_()
    k = torch.randn(seqlen, nhead, headdim, device=device, dtype=dtype).requires_grad_()
    v = torch.randn(seqlen, nhead, headdim, device=device, dtype=dtype).requires_grad_()

    attn_out, attn_lse, _, _ = _flash_attn_varlen_forward(
        q=q,
        k=k,
        v=v,
        cu_seqlens_q=cu_seqlens,
        cu_seqlens_k=cu_seqlens,
        max_seqlen_q=max_seqlen,
        max_seqlen_k=max_seqlen,
        softmax_scale=headdim ** (-0.5),
        causal=True,
        dropout_p=0.0,
    )
    attn_out_ref, attn_lse_ref = reference_attention(q, k, v, cu_seqlens, causal=True)
    assert torch.allclose(attn_out, attn_out_ref, atol=1e-2, rtol=1e-2)
    print("attn_lse.shape", attn_lse.shape)
    print("attn_lse_ref.shape", attn_lse_ref.shape)
    print("attn_lse", attn_lse)
    print("attn_lse_ref", attn_lse_ref)
    assert torch.allclose(attn_lse, attn_lse_ref, atol=1e-2, rtol=1e-2)



if __name__ == "__main__":
    test_flash_attn_varlen_three_seq()
    print("Flash attention varlen test passed.")

```
